test/integration/taint_add_test.go

```diff
@@ -0,0 +1,167 @@
+			}) {
+				return false
+			}
+			return true
+		}, eventuallyTimeout, eventuallyInterval).Should(gomega.BeTrue())
+
+		gomega.Eventually(func() bool {
+			if err := util.ApproveSpokeClusterCSR(kubeClient, managedClusterName, time.Hour*24); err != nil {
+				return false
+			}
+			return true
+		}, eventuallyTimeout, eventuallyInterval).Should(gomega.BeTrue())
+
+		// The managed cluster is available, so taints is expected to be empty
+		gomega.Eventually(func() bool {
+			managedCluster, err := util.GetManagedCluster(clusterClient, managedClusterName)
+			if err != nil {
+				return false
+			}
+			if len(managedCluster.Spec.Taints) != 0 {
+				return false
+			}
+			return true
+		}, eventuallyTimeout, eventuallyInterval).Should(gomega.BeTrue())
+
+		// make sure the managed available condition is cluster unknown
+		stop()
+
+		gomega.Eventually(func() bool {
+			managedCluster, err := util.GetManagedCluster(clusterClient, managedClusterName)
+			if err != nil {
+				return false
+			}
+			if len(managedCluster.Spec.Taints) != 1 {
+				return false
+			}
+			if managedCluster.Spec.Taints[0] != (v1.Taint{
+				Key:       "cluster.open-cluster-management.io/unreachable",
+				Value:     "ManagedClusterConditionAvailableUnknown",
+				Effect:    "NoSelect",
+				TimeAdded: managedCluster.Spec.Taints[0].TimeAdded,
+			}) {
+				return false
+			}
+			return true
+		}, eventuallyTimeout, eventuallyInterval).Should(gomega.BeTrue())
+
+		// restart the registration agent
+		go func() {
+			agentOptions := spoke.SpokeAgentOptions{
+				ClusterName:              managedClusterName,
+				BootstrapKubeconfig:      bootstrapKubeConfigFile,
+				HubKubeconfigSecret:      hubKubeconfigSecret,
+				HubKubeconfigDir:         hubKubeconfigDir,
+				ClusterHealthCheckPeriod: 1 * time.Minute,
+			}
+			err := agentOptions.RunSpokeAgent(context.Background(), &controllercmd.ControllerContext{
+				KubeConfig:    spokeCfg,
+				EventRecorder: util.NewIntegrationTestEventRecorder("cluster-tainttest"),
+			})
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		}()
+
+		// after one grace period, make sure the managed cluster available condition is recovered
+		// update available condition status manually
+		select {
+		case <-time.After(time.Duration(5*util.TestLeaseDurationSeconds+1) * time.Second):
+			managedCluster, err := util.GetManagedCluster(clusterClient, managedClusterName)
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			condition := meta.FindStatusCondition(managedCluster.Status.Conditions, v1.ManagedClusterConditionAvailable)
+			gomega.Expect(condition).ShouldNot(gomega.BeNil())
+			gomega.Expect(condition.Status).Should(gomega.Equal(metav1.ConditionTrue))
+			if condition != nil {
+				condition.Status = "False"
+			}
+			meta.SetStatusCondition(&(managedCluster.Status.Conditions), *condition)
+			managedCluster, err = clusterClient.ClusterV1().ManagedClusters().UpdateStatus(context.Background(), managedCluster, metav1.UpdateOptions{})
```

Comment:
Why we need to restart the agent?

Reply:
During the test, I found that if we do not restart the agent, the condition status may be changed to unknown automatically.